// homebridge-unifi-protect/dist/ffmpeg/protect-ffmpeg-exec.js
/* Copyright(C) 2017-2024, HJD (https://github.com/hjdhjd). All rights reserved.
 *
 * protect-ffmpeg-exec.ts: Execute arbitrary FFmpeg commands and return the results.
 *
 */
import { FfmpegProcess } from "./protect-ffmpeg.js";
export class FfmpegExec extends FfmpegProcess {
    isLoggingErrors;
    constructor(protectCamera, commandLineArgs, logErrors = true) {
        // Initialize our parent.
        super(protectCamera, commandLineArgs);
        // We want to log errors when they occur.
        this.isLoggingErrors = logErrors;
    }
    // Run the FFmpeg process and return the result.
    async exec(stdinData) {
        return new Promise((resolve) => {
            this.start();
            if (this.process === null) {
                this.log.error("Unable to execute command.");
                return null;
            }
            // Write data to stdin and close
            if (stdinData) {
                this.process.stdin.end(stdinData);
            }
            const stderr = [];
            const stdout = [];
            // Read standard output and standard error into buffers.
            this.process.stderr.on("data", (chunk) => stderr.push(chunk));
            this.process.stdout.on("data", (chunk) => stdout.push(chunk));
            // We prepend this listener to ensure we can properly cleanup after ourselves.
            this.process.prependOnceListener("exit", () => {
                // Trigger our process cleanup activities.
                this.stop();
            });
            // Return when process is done.
            this.process.once("exit", (exitCode) => {
                // Return the output and results.
                resolve({
                    exitCode,
                    stderr: Buffer.concat(stderr),
                    stdout: Buffer.concat(stdout)
                });
            });
        });
    }
    // Log errors.
    logFfmpegError(exitCode, signal) {
        // If we're ignoring errors, we're done.
        if (!this.isLoggingErrors) {
            return;
        }
        // Otherwise, revert to our default logging in our parent.
        super.logFfmpegError(exitCode, signal);
    }
}
